### PR TITLE
Multi-Pass Partitioning Decision (Multi-Pass PD)

### LINKED ISSUE #859: Reduce the number of block(s) tested @ MD
Design a fast/safe partitioning decision method towards a faster MD.

### PR DESCRIPTION
**Pull request by hguermaz**
## Description

Multi-Pass Partitioning Depth (Multi-Pass PD) performs multiple PD stages for the same SB towards 1 final Partitioning Structure. As we go from PDn to PDn + 1, the prediction accuracy of the MD feature(s) increases while the number of block(s) decreases.

For each SB, all blocks are tested in PD0 (4421 blocks if 128x128 SB, and 1101 blocks if 64x64 SB). Then the PD0 predicted Partitioning Structure is refined by considering up to three refinements depths away from the predicted depth, both in the direction of smaller block sizes and in the direction of larger block sizes (up to Pred - 3 / Pred + 3 refinement). The selection of the refinement depth is performed using the cost deviation between the current depth cost and candidate depth cost. The generated blocks are used as input candidates to PD1. The PD1 predicted Partitioning Structure is also refined (up to Pred - 1 / Pred + 1 refinement) using the square (SQ) vs. non-square (NSQ) decision(s) inside the predicted depth and using coefficient information. The final set of blocks is evaluated in PD2 to output the final Partitioning Structure.

**PD Stage(s) Setting(s)**

- PD0 MD Tool(s) : Best ME candidate only as INTER candidate(s), DC only as INTRA candidate(s), Chroma blind, Spatial SSE, no MVP table generation, no fast rate @ full cost derivation, Md-Stage 0 and Md-Stage 2 using count=1.
- PD1 MD Tool(s) : ME and Predictive ME only as INTER candidate(s) but MRP blind (only reference index 0 for motion compensation), DC only as INTRA candidate(s).
- PD2 MD Tool(s): default MD Tool(s).

**Available Mode(s)**
Added the ability to switch between 4 different multi-pass-pd modes @ picture basis through the existing signal pic_depth_mode:

- PIC_MULTI_PASS_PD_MODE_0 = 0: Multi-Pass PD Mode 0: PD0 | PD0_REFINEMENT.
- PIC_MULTI_PASS_PD_MODE_1 = 1: Multi-Pass PD Mode 1: PD0 | PD0_REFINEMENT | PD1 | PD1_REFINEMENT using SQ vs. NSQ only.
- PIC_MULTI_PASS_PD_MODE_2 = 2: Multi-Pass PD Mode 2: PD0 | PD0_REFINEMENT | PD1 | PD1_REFINEMENT using SQ vs. NSQ and SQ coeff info.
- PIC_MULTI_PASS_PD_MODE_3 = 3: Multi-Pass PD Mode 3: PD0 | PD0_REFINEMENT | PD1 | PD1_REFINEMENT using SQ vs. NSQ and both SQ and NSQ coeff info.

Closes #859 

P.S. This PR undo the Adaptive Open Loop Partitioning #788 changes.

## Author(s)

@hguermaz 
@PhoenixWorthVCD 

## Type of change

Enhancement

## Tests and performance
8BIT data generated for full-list objective-1-fast dataset in context of enc-mode 0. Speed data is generated on the 720p 8BIT clips for 5QP in term of cycles on AWS ec2 48 vCPU instances.

| pic_depth_mode| Speed Deviation| BD-Rate  Deviation|
| -- | -- | --
|PIC_MULTI_PASS_PD_MODE_0 | 13.70% | -0.004%|
|PIC_MULTI_PASS_PD_MODE_1 | 32.20% | 0.126% |
|PIC_MULTI_PASS_PD_MODE_2 | 49.90% | 0.329% | 
|PIC_MULTI_PASS_PD_MODE_3 | 51.20% | 0.433% | 

10BIT data generated for a set of 14 720p 10BIT clips in context of enc-mode 0. Speed data is generated on the same 14 720p 10BIT clips for 5QP in term of cycles on AWS ec2 48 vCPU instances.

| pic_depth_mode| Speed Deviation| BD-Rate  Deviation|
| -- | -- | --
|PIC_MULTI_PASS_PD_MODE_0 | 5.34% | -0.002%|
|PIC_MULTI_PASS_PD_MODE_1 | 8.11% | -0.003% |
|PIC_MULTI_PASS_PD_MODE_2 | 24.88% | 0.277% | 
|PIC_MULTI_PASS_PD_MODE_3 | 23.14% | 0.342% | 

Evaluation performed in context of QPS/QPM OFF.

### CODE REVIEW

File: Source/Lib/Common/Codec/EbEncDecProcess.c [hunk 1]
```diff
@@ -1686,6 +2126,455 @@ void move_cu_data(
...
+                                break;
+
+                            case BLOCK_8X8:
+                                zero_coeff_present_flag = (context_ptr->md_local_cu_unit[blk_index].avail_blk_flag && context_ptr->md_cu_arr_nsq[blk_index].block_has_coeff == 0) || // SQ
+                                    (context_ptr->md_local_cu_unit[blk_index + 1].avail_blk_flag && context_ptr->md_cu_arr_nsq[blk_index + 1].block_has_coeff == 0 && context_ptr->md_cu_arr_nsq[blk_index + 2].block_has_coeff == 0) || // H
+                                    (context_ptr->md_local_cu_unit[blk_index + 3].avail_blk_flag && context_ptr->md_cu_arr_nsq[blk_index + 3].block_has_coeff == 0 && context_ptr->md_cu_arr_nsq[blk_index + 4].block_has_coeff == 0);  // V
+                                break;
+
+                            case BLOCK_4X4:
+                                zero_coeff_present_flag = (context_ptr->md_cu_arr_nsq[blk_index].block_has_coeff == 0); // SQ
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbEncDecProcess.c]: there is a way to change this switch case to a for loop for all cases

File: Source/Lib/Common/Codec/EbModeDecision.c [hunk 2]
```diff
@@ -987,7 +1030,10 @@ void Unipred3x3CandidatesInjection(
         const MeCandidate *me_block_results_ptr = &me_block_results[me_candidate_index];
         const uint8_t inter_direction = me_block_results_ptr->direction;
         const uint8_t list0_ref_index = me_block_results_ptr->ref_idx_l0;
-
+#if MULTI_PASS_PD
+        if (list0_ref_index > context_ptr->md_max_ref_count - 1)
+            continue;
+#endif
```
Comment (chkngit) [inline Source/Lib/Common/Codec/EbModeDecision.c]: thanks

File: Source/Lib/Common/Codec/EbFullLoop.c [hunk 3]
```diff
@@ -2019,11 +2022,19 @@ void product_full_loop_tx_search(
...
+#if MULTI_PASS_PD
+    if (context_ptr->tx_search_reduced_set == 2)
+#else
     if (picture_control_set_ptr->parent_pcs_ptr->tx_search_reduced_set == 2)
+#endif
         txk_end = 2;
 
     for (int32_t tx_type_index = txk_start; tx_type_index < txk_end; ++tx_type_index) {
+#if MULTI_PASS_PD
+        if (context_ptr->tx_search_reduced_set == 2)
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbFullLoop.c]: Just a comment, would be nice if this was refactored to use ENUMs instead.

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 4]
```diff
@@ -32,6 +32,9 @@
 #ifdef __cplusplus
 extern "C" {
 #endif
+
+#define MULTI_PASS_PD                1
```
Comment (mafonsoN) [inline Source/Lib/Common/Codec/EbDefinitions.h]: Can you please add a comment next to each new #define ?

File: Source/Lib/Common/Codec/EbFullLoop.c (hunk 3 above)
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbFullLoop.c]: I agree. Please note that here we just moved tx_search_reduced_set from Parent PCS data structure to EncDec context data structure as: (1) only EncDec uses tx_search_reduced_set, (2) to control tx_search_reduced_set from a PD stage to another. We will make sure to address this (and more) during the clean up phase.

File: Source/Lib/Common/Codec/EbRateControlProcess.c [hunk 5]
```diff
@@ -3229,7 +3229,7 @@ enum {
 #define LOW_QPS_COMP_THRESHOLD         40
 #define HIGH_FILTERED_THRESHOLD     (4<<8) // 8 bit precision
 #define LOW_FILTERED_THRESHOLD      (2<<8) // 8 bit precision
-#define QPS_SW_THRESH          8
+#define QPS_SW_THRESH                  100 // 8
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbRateControlProcess.c]: This change is made to shut QPS and QP. In fact the default mode should be CORE only (the evaluation is performed in context of CORE). An upcoming PR is going to take care of controlling QPS and QP trough an API signal.

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 4 above)
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Done

**Comment by hguermaz** [comment]
> What is the performance expected for the default settings on enc_mode 0?
@mafonsoN Defaulrt mode for M0 is PIC_MULTI_PASS_PD_MODE_1.


8BIT: ~32.20% speed gain and  0.126%

File: Source/Lib/Common/Codec/EbRateDistortionCost.c [hunk 6]
```diff
@@ -796,8 +802,14 @@ uint64_t av1_intra_fast_cost(
     chromaRate = (uint32_t)(intraChromaModeBitsNum + intraChromaAngModeBitsNum);
 
     // Keep the Fast Luma and Chroma rate for future use
+
```
[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbRateDistortionCost.c]: Done

File: Source/Lib/Common/Codec/EbFullLoop.c (hunk 3 above)
Comment (mafonsoN) [reply Source/Lib/Common/Codec/EbFullLoop.c]: Thanks @hguermaz , we need to make sure we add these to the list of items to cleanup then.

**Comment by mafonsoN** [comment]
> > What is the performance expected for the default settings on enc_mode 0?
> > @mafonsoN Defaulrt mode for M0 is PIC_MULTI_PASS_PD_MODE_1.
> 
> 8BIT: ~32.20% speed gain and 0.126%

Okay, for your results, the depth mode of the non I-slices is changed, and the I-slices always use PIC_ALL_DEPTH_MODE, right? Also, it seems like PIC_MULTI_PASS_PD_MODE_0 is not used for any mode.

File: Source/Lib/Common/Codec/EbRateControlProcess.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (hassount) [reply Source/Lib/Common/Codec/EbRateControlProcess.c]: I believe this should not change yet until we introduce the qpm qps signalling

[the remark this answers is not in the document]
Comment (hguermaz) [reply Source/Lib/Common/Codec/EbRateControlProcess.c]: @hassount sure. Set back default value to 8.

**Comment by hassount** [comment]
@mafonsoN 
have all points been addressed ? can we merge as soon as checks are passing ?

**Comment by mafonsoN** [comment]
> @mafonsoN
> have all points been addressed ? can we merge as soon as checks are passing ?

Yes

**Comment by grill05** [comment]
> What is the performance expected for the default settings on enc_mode 0?
> @mafonsoN Default mode for M0 is PIC_MULTI_PASS_PD_MODE_1.
> 
> 8BIT: ~32.20% speed gain and 0.126%

@hguermaz @PhoenixWorthVCD Just trying to understand the naming convention here. 
Do PIC_MULTI_PASS_PD_MODE_1,PIC_MULTI_PASS_PD_MODE_2,PIC_MULTI_PASS_PD_MODE_3,PIC_MULTI_PASS_PD_MODE_4 etc get enabled at M0,M1,M2,M3 ?

**Comment by hguermaz** [comment]
> > What is the performance expected for the default settings on enc_mode 0?
> > @mafonsoN Default mode for M0 is PIC_MULTI_PASS_PD_MODE_1.
> > 8BIT: ~32.20% speed gain and 0.126%
> 
> @hguermaz @PhoenixWorthVCD Just trying to understand the naming convention here.
> Do PIC_MULTI_PASS_PD_MODE_1,PIC_MULTI_PASS_PD_MODE_2,PIC_MULTI_PASS_PD_MODE_3,PIC_MULTI_PASS_PD_MODE_4 etc get enabled at M0,M1,M2,M3 ?

@grill05 PIC_MULTI_PASS_PD_MODE_0,1,2,3 are four distinct levels for the same feature: Multi-Pass PD. Multi-Pass PD provides better Speed/BD-Rate trade-offs than PIC_ALL_DEPTH_MODE (faster and reasonable BD-Rate losses). Note that this feature is not yet used for I_SLICE only (but supported). M0 deploys PIC_MULTI_PASS_PD_MODE_1, while M1 & M2 deploys PIC_MULTI_PASS_PD_MODE_2. Multi-Pass PD is currently not used for M1 & beyond. Also please note that PIC_MULTI_PASS_PD_MODE_0 and PIC_MULTI_PASS_PD_MODE_3 are supported, but not yet used.

**Pull request merged**